### PR TITLE
Auto-tested CRUD on Accounts & minor fixes

### LINKED ISSUE #2: AccountsController.cs Implementation
as in the title

### LINKED ISSUE #3: Accounts Controller Integration Tests
as in the title

### LINKED ISSUE #4: Add Swagger UI
it's just useful while debugging

### LINKED ISSUE #5: Update NuGet Packages
old EF (used in this project) has a weird issue with VS Developers' PowerShell; just need to update;

### PR DESCRIPTION
**Pull request by Bazik33**
Fixes #2 
Fixes #3 
Fixes #4 
Fixes #5

### CODE REVIEW

File: backend/src/NotePaste.WebApi/Controllers/AccountsController.cs [hunk 1]
```diff
@@ -12,44 +24,325 @@ public class AccountsController : Controller
...
+        private SmtpClient _client;
+        private FrontendSettings _frontendSettings;
 
-        public AccountsController(UserManager<AppUser> userManager, RoleManager<IdentityRole> roleManager)
+        public AccountsController(UserManager<AppUser> userManager, RoleManager<IdentityRole> roleManager, EmailSettings emailSettings, FrontendSettings frontendSettings)
         {
             _userManager = userManager;
             _roleManager = roleManager;
+            _emailSettings = emailSettings;
+            _client = new SmtpClient();
```
Comment (KanarekLife) [inline backend/src/NotePaste.WebApi/Controllers/AccountsController.cs:36]: Why did you add SmtpClient as singleton when you initialize it here? 
You should be it as transitend for example in order to make it possible to unit test if emails are being sent correctly.

File: backend/src/NotePaste.WebApi/Controllers/AccountsController.cs [hunk 2]
```diff
@@ -12,44 +24,325 @@ public class AccountsController : Controller
...
+            }
+            else
+            {
+                return BadRequest();
+            }
+        }
+
         /// <summary>
-        /// Register
+        /// Creates new user
```
Comment (KanarekLife) [inline backend/src/NotePaste.WebApi/Controllers/AccountsController.cs:72]: If you change that into more CRUD approach then change method name to CreateUser

File: backend/src/NotePaste.WebApi/Controllers/AccountsController.cs [hunk 3]
```diff
@@ -12,44 +24,325 @@ public class AccountsController : Controller
...
         /// </summary>
         /// <param name="model"></param>
         /// <returns>An info whether registration was successful</returns>
+        /// <response code="200">When registration is completed successfully </response>
```
Comment (KanarekLife) [inline backend/src/NotePaste.WebApi/Controllers/AccountsController.cs:76]: Better code for created CRUD content is 201

File: backend/src/NotePaste.WebApi/Controllers/AccountsController.cs [hunk 4]
```diff
@@ -12,44 +24,325 @@ public class AccountsController : Controller
...
+        [ProducesResponseType(StatusCodes.Status409Conflict)]
         public async Task<IActionResult> Register(RegisterModel model)
         {
             var user = await _userManager.FindByNameAsync(model.Username) ?? await _userManager.FindByEmailAsync(model.Email);
             if (user != null)
             {
-                return Conflict();
+                if (user.UserName == model.Username)
+                {
+                    return Conflict("Username is already taken");
```
Comment (KanarekLife) [inline backend/src/NotePaste.WebApi/Controllers/AccountsController.cs:90]: Why are you adding explenation in api responses when you have docs for that?

File: backend/src/NotePaste.WebApi/Controllers/AccountsController.cs [hunk 5]
```diff
@@ -12,44 +24,325 @@ public class AccountsController : Controller
...
+                }
+                if (user.Email == model.Email)
+                {
+                    return Conflict("An account with this email address is already created");
+                }
+            }
+
+                string pattern = @"^(?!\.)(""([^""\r\\]|\\[""\r\\])*""|([-a-z0-9!#$%&'*+/=?^_`{|}~]|(?<!\.)\.)*)(?<!\.)@[a-z0-9][\w\.-]*[a-z0-9]\.[a-z][a-z\.]*[a-z]$";
```
Comment (KanarekLife) [inline backend/src/NotePaste.WebApi/Controllers/AccountsController.cs:98]: `Another way is to use the System.Net.Mail.MailAddress class. To determine whether an email address is valid, pass the email address to the MailAddress.MailAddress(String) class constructor.`

File: backend/src/NotePaste.WebApi/Controllers/AccountsController.cs [hunk 6]
```diff
@@ -12,44 +24,325 @@ public class AccountsController : Controller
...
+
+                if (result.Succeeded)
+                {
+                    var token = await _userManager.GenerateEmailConfirmationTokenAsync(user);
+                    await SendConfirmationEmail(user.Email, token);
+                    return Ok();
+                }
+                else
+                {
+                    return BadRequest(result.Errors);
```
Comment (KanarekLife) [inline backend/src/NotePaste.WebApi/Controllers/AccountsController.cs:121]: Have you added how ther errors response will look like in docs (swagger)?

File: backend/src/NotePaste.WebApi/Controllers/AccountsController.cs [hunk 7]
```diff
@@ -12,44 +24,325 @@ public class AccountsController : Controller
...
             }
-            
-            user = new AppUser()
+            else
             {
-                Email = model.Email,
-                UserName = model.Username
-            };
-            var result = await _userManager.CreateAsync(user, model.Password);
+                return BadRequest();
```
Comment (KanarekLife) [inline backend/src/NotePaste.WebApi/Controllers/AccountsController.cs:193]: Should be 500 Internal Error

File: backend/src/NotePaste.WebApi/Controllers/AccountsController.cs [hunk 8]
```diff
@@ -12,44 +24,325 @@ public class AccountsController : Controller
...
                 return Ok();
             }
             else
             {
-                return BadRequest(result.Errors);
+                return BadRequest();
             }
         }
+
+        private async Task SendConfirmationEmail(string email, string token)
```
Comment (KanarekLife) [inline backend/src/NotePaste.WebApi/Controllers/AccountsController.cs:293]: These methods should not be in controller.
These methods should be in separete service that is heavily unit tests

File: backend/src/NotePaste.WebApi/Controllers/AccountsController.cs [hunk 9]
```diff
@@ -12,44 +24,325 @@ public class AccountsController : Controller
...
+        {
+            string url = _frontendSettings.EmailConfirmationUrl + "?token=" + token;
+            string subject = "Email confirmation on NotePaste";
+            string content = "<!DOCTYPE html> <html lang=\"en-us\"> <head> <meta charset=\"utf-8\" /> </head> <body>" +
```
Comment (KanarekLife) [inline backend/src/NotePaste.WebApi/Controllers/AccountsController.cs:297]: Why are you defining Html content in mail?

File: backend/src/NotePaste.WebApi/Controllers/AccountsController.cs [hunk 10]
```diff
@@ -1,8 +1,20 @@
...
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Http;
 using Microsoft.AspNetCore.Identity;
 using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore.Metadata.Internal;
+using MimeKit;
 using NotePaste.Core.Entities;
 using NotePaste.WebApi.Models.AccountModels;
+using NotePaste.WebApi.Models.SettingsModels;
+using Renci.SshNet;
```
[the remark this answers is not in the document]
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/Controllers/AccountsController.cs:17]:
```suggestion
using Renci.SshNet;
```

mistake, I copied it with a snippet from Stackoverflow

File: backend/src/NotePaste.WebApi/Controllers/AuthController.cs [hunk 11]
```diff
@@ -80,6 +80,7 @@ public async Task<IActionResult> Login(LoginModel model)
         /// <response code="200">JWT token is correct</response>
         [Route("/amiauthorized")]
         [Authorize]
+        [HttpGet]
```
[the remark this answers is not in the document]
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/Controllers/AuthController.cs:83]: We need to point out that it is GET because of Swagger (it crashes with ambiguous Method). But it's a good point to remove [Route] & [HttpGet] and put `[HttpGet("/amiauthorized")]`

```suggestion
        [HttpGet("/amiauthorized")]
```

File: backend/src/NotePaste.Core/Entities/User.cs [hunk 12]
```diff
@@ -5,6 +5,6 @@ namespace NotePaste.Core.Entities
 {
     public class AppUser : IdentityUser
     {
-        
+        public string PhotoUrl { get; set; }
```
[the remark this answers is not in the document]
Comment (Bazik33) [reply backend/src/NotePaste.Core/Entities/User.cs:8]:
```suggestion
        [MaxLength(256)]
        public string PhotoUrl { get; set; }
```

File: backend/src/NotePaste.WebApi/Migrations/AppDbContextModelSnapshot.cs [hunk 13]
```diff
@@ -187,6 +187,9 @@ protected override void BuildModel(ModelBuilder modelBuilder)
                     b.Property<bool>("PhoneNumberConfirmed")
                         .HasColumnType("bit");
 
+                    b.Property<string>("PhotoUrl")
+                        .HasColumnType("text");
```
[the remark this answers is not in the document]
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/Migrations/AppDbContextModelSnapshot.cs:191]: Fixed in latest commit.

File: backend/src/NotePaste.WebApi/Migrations/20200509181056_05092020-0809pm-NewProperties.cs [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+namespace NotePaste.WebApi.Migrations
+{
+    public partial class _050920200809pmNewProperties : Migration
+    {
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.AddColumn<string>(
+                name: "PhotoUrl",
+                table: "AspNetUsers",
+                nullable: true);
```
[the remark this answers is not in the document]
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/Migrations/20200509181056_05092020-0809pm-NewProperties.cs:12]: Fixed in latest commit.

File: backend/src/NotePaste.WebApi/Controllers/AuthController.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/Controllers/AuthController.cs:83]: Fixed in latest commit.

File: backend/src/NotePaste.WebApi/Models/AccountModels/ConfirmEmailModel.cs [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+using System.Threading.Tasks;
+
+namespace NotePaste.WebApi.Models.AccountModels
+{
+    public class ConfirmEmailModel
+    {
+        [Required]
+        [MaxLength(128)]
+        [JsonPropertyName("email")]
+        public string email { get; set; }
```
[the remark this answers is not in the document]
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/Models/AccountModels/ConfirmEmailModel.cs:15]:
```suggestion
        [Required]
        [MaxLength(256)]
        [JsonPropertyName("email")]
        public string email { get; set; }
```

[the remark this answers is not in the document]
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/Models/AccountModels/ConfirmEmailModel.cs:15]: Fixed in latest commit.

File: backend/src/NotePaste.WebApi/Models/AccountModels/DeleteModel.cs [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+using System.Threading.Tasks;
+
+namespace NotePaste.WebApi.Models.AccountModels
+{
+    public class DeleteModel
+    {
+        [Required]
+        [MaxLength(128)]
+        [JsonPropertyName("password")]
+        public string password { get; set; }
```
[the remark this answers is not in the document]
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/Models/AccountModels/DeleteModel.cs:15]: Fixed in latest commit.

File: backend/src/NotePaste.WebApi/Models/AccountModels/ResetPasswordModel.cs [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+using System.Text.Json.Serialization;
+using System.Threading.Tasks;
+
+namespace NotePaste.WebApi.Models.AccountModels
+{
+    public class ResetPasswordModel
+    {
+        //string email, string token, string newPassword
```
[the remark this answers is not in the document]
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/Models/AccountModels/ResetPasswordModel.cs:12]:
```suggestion
```

There is no point, copied by accident.

File: backend/src/NotePaste.WebApi/Controllers/AccountsController.cs (hunk 7 above)
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/Controllers/AccountsController.cs:192]:
```suggestion
                return  StatusCode(StatusCodes.Status500InternalServerError); 
```

File: backend/src/NotePaste.WebApi/Controllers/AccountsController.cs (hunk 9 above)
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/Controllers/AccountsController.cs:296]: Because we're writing emails in HTML.

File: backend/src/NotePaste.WebApi/NotePaste.WebApi.csproj [hunk 18]
```diff
@@ -1,24 +1,34 @@
-<Project Sdk="Microsoft.NET.Sdk.Web">
+﻿<Project Sdk="Microsoft.NET.Sdk.Web">
 
     <PropertyGroup>
         <TargetFramework>netcoreapp3.1</TargetFramework>
     </PropertyGroup>
 
+    <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
+      <DocumentationFile>C:\Users\bazik\Source\Repos\Bazik33\NotePaste\backend\src\NotePaste.WebApi\NotePaste.WebApi.xml</DocumentationFile>
```
[the remark this answers is not in the document]
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/NotePaste.WebApi.csproj]: Fixed.

File: backend/src/NotePaste.WebApi/NotePaste.WebApi.xml [hunk 19]
```diff
@@ -0,0 +1,91 @@
+<?xml version="1.0"?>
```
[the remark this answers is not in the document]
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/NotePaste.WebApi.xml:1]: Added.

File: backend/tests/NotePaste.IntegrationTests/AccountsController/RegisterTests.cs [hunk 20]
```diff
@@ -49,6 +49,70 @@ public async void PreventRegistrationWhenPasswordTooWeak()
...
+
+            var response = await _client.PostAsync("/accounts/register", await Helpers.CreateJsonContent(model));
+            var responseContent = await response.Content.ReadAsStringAsync();
+            Assert.Equal(HttpStatusCode.BadRequest, response.StatusCode);
+        }
+
+        [Fact]
+        public async void PreventRegisterWhenUsernameIsAlreadyTaken()
+        {
+            var model = new RegisterModel()
```
[the remark this answers is not in the document]
Comment (Bazik33) [reply backend/tests/NotePaste.IntegrationTests/AccountsController/RegisterTests.cs:70]: fixed in last commit.

File: backend/src/NotePaste.WebApi/Controllers/AccountsController.cs (hunk 4 above)
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/Controllers/AccountsController.cs:89]: I think we need to inform the frontend whether username OR password is already taken.

File: backend/src/NotePaste.WebApi/Controllers/AccountsController.cs (hunk 2 above)
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/Controllers/AccountsController.cs:71]: Fixed in last commit.

File: backend/src/NotePaste.WebApi/Controllers/AccountsController.cs (hunk 3 above)
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/Controllers/AccountsController.cs:75]: Fixed in last commit.

File: backend/src/NotePaste.WebApi/Startup.cs [hunk 21]
```diff
@@ -29,15 +34,38 @@ public void ConfigureServices(IServiceCollection services)
...
             services.AddJwtAuthenticationServices(_configuration);
             services.AddCors();
+
+            var emailSettings = new EmailSettings();
+            _configuration.Bind("EmailSettings", emailSettings);
+            services.AddSingleton(emailSettings);
+            var frontendSettings = new FrontendSettings();
+            _configuration.Bind("FrontendSettings", frontendSettings);
+            services.AddSingleton(frontendSettings);
+            services.AddSingleton(new SmtpClient());
```
[the remark this answers is not in the document]
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/Startup.cs:55]: Deleted SmtpClient singleton.

File: backend/src/NotePaste.WebApi/Controllers/AccountsController.cs (hunk 1 above)
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/Controllers/AccountsController.cs:35]: Fixed thru 'Deleted SmtpClient singleton'

File: backend/src/NotePaste.WebApi/Controllers/AccountsController.cs (hunk 5 above)
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/Controllers/AccountsController.cs:97]: Changed.

File: backend/src/NotePaste.WebApi/Startup.cs [hunk 22]
```diff
@@ -29,15 +34,38 @@ public void ConfigureServices(IServiceCollection services)
...
             {
                 options.UseMySQL(_configuration.GetConnectionString("Default"), b => b.MigrationsAssembly("NotePaste.WebApi"));
             });
+
             services.AddHealthChecks()
                 .AddDbContextCheck<AppDbContext>("ApplicationDbContext");
             services.AddJwtAuthenticationServices(_configuration);
             services.AddCors();
+
+            var emailSettings = new EmailSettings();
```
[the remark this answers is not in the document]
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/Startup.cs:49]: Done.

File: backend/src/NotePaste.WebApi/Controllers/AccountsController.cs (hunk 6 above)
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/Controllers/AccountsController.cs:119]: Do you mean this? https://swagger.io/docs/specification/describing-responses/

File: backend/src/NotePaste.WebApi/Controllers/AccountsController.cs (hunk 8 above)
Comment (Bazik33) [reply backend/src/NotePaste.WebApi/Controllers/AccountsController.cs:291]: I agree, I'll split it and write unit tests.

File: backend/tests/NotePaste.IntegrationTests/AccountsController/Update_Fetch_Delete_Test.cs [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+        public Update_Fetch_Delete_Test(TestWebApplication factory)
+        {
+            _client = factory.CreateClient();
+        }
+
+        [Fact]
+        public async void CRUD_WithPrecreatedUser()
+        {
+            // Get a JWT
+            var authModel = new LoginModel()
```
[the remark this answers is not in the document]
Comment (Bazik33) [reply backend/tests/NotePaste.IntegrationTests/AccountsController/Update_Fetch_Delete_Test.cs:33]: I'll do this soon.

File: backend/src/NotePaste.WebApi/Controllers/AccountsController.cs (hunk 6 above)
Comment (KanarekLife) [reply backend/src/NotePaste.WebApi/Controllers/AccountsController.cs:119]: For example. I personally thought about creating some sort of response class for this controller and just use it but it's also a good way to go i think.
